### PR TITLE
azure - add trim tag action 

### PR DESCRIPTION
**Pull request by erwelch**

### CODE REVIEW

File: tools/c7n_azure/tests/test_tags.py [hunk 1]
```diff
@@ -424,4 +427,204 @@ def test_auto_tag_days_must_be_btwn_1_and_90(self):
...
+            'actions': [
+                {'type': 'tag-trim',
+                 'space': 0,
+                 'preserve': ['testtag']
+                 }
+            ],
+        })
+        p.run()
+
+        logger_mock.assert_called()
```
Comment (aluong) [inline tools/c7n_azure/tests/test_tags.py]: Is it possible to get a more detailed assert; I don't want this test to have false positives? I suppose the patch for 'warning' might be sufficient.

File: tools/c7n_azure/c7n_azure/actions.py [hunk 2]
```diff
@@ -243,3 +242,83 @@ def get_first_operation(logs, operation_name):
...
+    def process(self, resources):
+        preserve = set(self.data.get('preserve', {}))
+        space = self.data.get('space')
+        untag_action = self.manager.action_registry.get('untag')
+
+        for resource in resources:
+            # get existing tags
+            tags = resource.get('tags', {})
+
+            # space == 0 means remove all but specified
```
[the remark this answers is not in the document]
Comment (erwelch) [reply tools/c7n_azure/c7n_azure/actions.py]: Yep, absolutely! I am considering removing the comment and just adding it as part of the description.

File: tools/c7n_azure/tests/test_tags.py (hunk 1 above)
Comment (erwelch) [reply tools/c7n_azure/tests/test_tags.py]: Yeah I could do called_with but it's going to be called with a long interpolated string with the resource's id, but I'll see if I can make it more robust.

File: tools/c7n_azure/c7n_azure/actions.py [hunk 3]
```diff
@@ -243,3 +242,83 @@ def get_first_operation(logs, operation_name):
...
+      - policies:
+         - name: azure-tag-trim
+           comment: |
+             Any instances with 8 or more tags get tags removed until
+             they match the target tag count, in this case 7 so
+             that we free up tag slots for another usage.
+           resource: azure.resourcegroup
+           filters:
+               # Filter down to resources which already have 8 tags
+               # as we need space for 3 more, this also ensures that
```
[the remark this answers is not in the document]
Comment (erwelch) [reply tools/c7n_azure/c7n_azure/actions.py]: Yeah I copied it from aws. It's essentially saying to make the action more performant you can filter down to the resources you know already have too many tags that need trimming. I'll try and come up with something  easier to understand.

File: tools/c7n_azure/c7n_azure/actions.py [hunk 4]
```diff
@@ -243,3 +242,83 @@ def get_first_operation(logs, operation_name):
...
+    """
+    max_tag_count = 15
+
+    schema = utils.type_schema(
+        'tag-trim',
+        space={'type': 'integer'},
+        preserve={'type': 'array', 'items': {'type': 'string'}})
+
+    def validate(self):
+        if (self.data.get('space') is not None and
```
[the remark this answers is not in the document]
Comment (erwelch) [reply tools/c7n_azure/c7n_azure/actions.py]: TIL None < 0 evaluates to True! So nope. Will remove.

**Pull request merged**